src/transformers/trainer.py

```diff
@@ -52,7 +52,7 @@
 import torch
 from packaging import version
 from torch import nn
-from torch.utils.data import DataLoader, Dataset, IterableDataset, RandomSampler, SequentialSampler
+from torch.utils.data import DataLoader, Dataset, RandomSampler, SequentialSampler
```

Comment:
@sgugger - could you take a look here as well? :-)

Reply:
Those changes seem to be needed to make Trainer + DDP with streaming work

Reply:
For context: `len(IterableDatasetShard())` raises an error (as expected) if its underlying dataset (`datasets.TorchIterableDataset` in our case) is not sized. So for a deeper `Sized` check we have to get to `IterableDatasetShard.dataset`, hence the `dataloader.dataset.dataset` below.

Reply:
I don't understand those changes. `IterableDatasetShard` [implements `__len__`](https://github.com/huggingface/transformers/blob/aa19f478acc8f46b7f70e9c525a2bd07dfed548c/src/transformers/trainer_pt_utils.py#L785) specifically for supporting iterable datasets with length (usually coming from a Dataset with straeming), with the method throwing an error when the underlying dataset does not implements `__len__`.

Or is it that `IterableDatasetShard` is always considered an `abc.collections.Sized` since there is the implementation of the method and you need to have them not considered sized in this case? If that's the use case, we should replace the check from `abc` with a try except to get the length.

Reply:
@sgugger yes, the second case is exactly what we have here! Wasn't sure how you feel about try:except blocks style-wise, but if it's ok I'll replace the checks :slightly_smiling_face:

Reply:
Ok so then can you add an util `has_length` that tries to access the length of the dataset and returns False in case of error, then use this instead of diving in three attributes? It'd be cleaner :-)